### ISSUE TITLE
[mnm] Use the Client Communicator to push notifications and management call answers (not stats) back to the TMS Client Entity

### ISSUE THREAD

**Issue by mathieucarbou** [Terracotta-OSS/terracotta-platform#195]
The goal of this change is to get the M&M data as fast as possible and with less data loss __before arriving inside TMS__

What I don't really like with the current architecture is that the m&m data is saved temporary inside voltrons in a limited buffer and we can hardly control it. Also, TMS needs to frequently pool each active to get the m&m data. So there is both a delay and a potential data loss BEFORE data arrives in TMS.

I'd like that the m&m data arrive in TMS __as fast as possible__ and if there are some concerns about buffering the data or things like that, that could cause some loss, then it's up to the TMS to control that.

Main idea is that the data arrive fast and that we can control how we handle it / process it / temporary store it in memory.

This would replace both things:

-  the fact that we need to poll from TMS
- the ring buffer that needs to act as a "temporary limited buffer" inside voltron

__To consider__

- TMS does not need to go through the traditional invoke() calls of the entities to get messages and do polling because we do not care about our calls to be replicated. Using client communicator is fine for us.

__Current Architecture__

1. Notification / Stats pushed inside monitoring service
2. Monitoring service put them inside each consumer’s ring buffer (_ring buffer inside voltron - not good and size difficult to change / control - linked to an entity_)
3. TMS -> polls -> TMS Entity -> consumes ring buffer (_delay and data miss__)

__Proposal__

1. Notification / Stats pushed inside monitoring service
2. Monitoring service uses the consumer’s client communicator to directly push these to the client (here, tms entity), by using the client communicator of the consumer
3. __TMS received asynchronously in real-time the data__
4. TMS put the data in its own ring-buffer (__size controlled within the app, data-loss not relative to voltron but TMS__)
5. In TMS, a consumer thread reads the ring buffer to process the data (__velocity and impacts are within the TMS now__)

Please share your comments :-)

**Comment by jd0-sag** [comment]
There are a few fundamental questions which need to be addressed with this.  First and foremost is whether or not M&M _should_ be doing what it can to get the most data as quickly as possible.  Remember that much of the desire (as I recall) is that it can throttle back before the application is significantly impacted.  Hence, be careful in attempts to avoid any kind of back-off caused by a busy server or saturated network.

Pushing directly from the service introduces a few issues:

- if you are pushing the message in response to the call from the platform to populate the information, you are doing this work on a thread which probably shouldn't:  in the active-only case, it is likely just being pushed down from the entity calling it, while in the passive case, it is likely being run on a thread managing the group channel.
- determining which client should receive the call will be racy since the client can connect/disconnect independently of the call into the service since they are not part of the same entity.

Now, you could use the timer invoke or entity messenger to move this call into your own thread in a normal invoke but, if you are serious about not dropping any data, you would be potentially queuing up lots of them (this would be safe but may demonstrate the issue with this approach).

I would have thought that the bigger interest in this was in being able to degrade as network or CPU saturation increased by being able to transition to less frequent polls of larger pieces of data (waiting on the invoke response would give you a natural back-pressure mechanism to implement this "less frequent poll" approach).

This is also why the monitoring interface was designed with 2 different kinds of data in mind:  high-fidelity (data structure events as mutations to allow blind collapsing of stale data, if the poll isn't keeping up) or low-fidelity (key-value data where successive writes to the same key can blindly overwrite the stale values).

Instead of forcing this into a fast and highly reliable channel, wasn't this one intentions to naturally degrade so it could still monitor a busy cluster without being the reason why it was busy?

I suppose the question of relative priority between M&M operations and application operations must be answered to determine the path forward.

**Comment by lorban** [comment]
I second @jd0-sag in pretty much everything he said: pushing the stats over the network 'in real-time' is unfeasible: for instance, each cache hit modifies a counter and would require a network call with such design. Even optimized to the bone, we're talking at least two orders of magnitude of difference between the counter shipping vs the actual cache work.

Monitoring theory categorizes some artifacts as natural poll candidates (like counters) and some others as natural push candidates (like events). Let's follow the best practices of the industry and not try to work around them.

**Comment by mathieucarbou** [comment]
Note to tall: I think we can then keep as-is for the statistics, I agree. But the question remains for:

- all notifications (platform events and application-level events)
- all management call returns.

Currently, in the management entity, we are __ALREADY__ using the client communicator to pass management call requests to the clients. I'd like to also setup a client communicator in the TMS Entity so that management call answers can be forwarded back to the TMS quickly. We do not want to wait 5 seconds in the TMS to receive a management call answer: it simply does not make sense. Also, I'd like to re-use the same mechanism to send notifications to the TMS (through client communicator). Because we currently loose some important notiifcations, or they do not arrive on time: in 5 seconds, a lot of things can happen in cluster.

__Reasons:__

- we are ALREADY using the client communicator (for client's mangementcalls)
- the tms needs to poll for messages. So currently, it means that we cannot have the answer to a management call BEFORE 5 seconds
- By this time, also, some IMPORTANT notifications can be lost.

I agree that for statistics it can be discussed and you're right, but for management call answers that comes asynchronously and for notifications also, if we do not get these objects through the client communicator, it means we need to store them in this ring buffer and they get polled. It means that the TMS can loose some IMPORTANT notifications and all management calls will be really SLOW because the TMS needs to wait for the next poll to get the answers.

We are already using the client communicator to send management calls to clients. So I suggest we extend its use:

1. we leave the ring buffer in voltron for stats (=> polling)
2. we use the client communicator ALSO to send management calls answers (like we already do for requests) to the TMS
3. we also use the client communicator to send important notifications to the TMS

These changes are really minor (2-3 hours of work I think). And it will I think be a good alternative between your concerns and what we need.

**Comment by mathieucarbou** [comment]
@anthonydahanne @GaryWKeim  @lorban @jd0-sag : Just to make things clear, when I said push to TMS the information as soon as we get it, for stats, does not mean that they will be pushed in 1 second (sampling frequency)!

How it works now:

__for stats__

1. stats are sampled each 1 second in a circular buffer
2. a collector thread collects them each 75% * Time To Disable (=> approx. each 22 second)
3. This collector sends the stats inside voltron's monitoring buffer
4. tms polls each 5 second

=> I want to shortcut step 3: instead of putting the stats inside a ring buffer, the tms entity can just send them back to TMS with the client communicator. => It's an operation which will be done each 22 seconds. Not each 1 second.

__ for notifs__

1. notifs is generated from a client, entity, etc.
2. notifs are pushed inside the voltron's monitoring buffer
3. tms polls each 5 second

=> I want to also shortcut step 2, to send back the notif as they arrive to the TMS's client communicator.

__ for management call on client or server entity (how it is supposed to work)__

1. TMS do a management call
2. management call request is sent to the client with client communicator, or the entity (with replication if necessary)
3. call is executed
4. response is sent back through the management entity (for clients) or through the monitoring producer (for entities)
5. then response is put inside this monitoring ring buffer
6. tms polls each 5 second, and get the call response

=> I want to shortcut step 5 so that we can get the management call answer as fast as possible with the client communicator.

**Comment by GaryWKeim** [comment]
That all seems perfectly reasonable.

**Comment by mathieucarbou** [comment]
@anthonydahanne @ljacomet @GaryWKeim : after meeting discussion.

- Tms Entity just needs to send itself a message each 1 second for example to dequeue de messages in the rung buffer and send them to the the tms client with the client communicator (=> `IEntityMessenger`)
- ringbuffer should discard stats messages first

__PAY ATTENTION TO_ not send client communicator messages from for example a push from ehcache client.

**Comment by mathieucarbou** [comment]
question to @ljacomet @jd0-sag @myronkscott : `IEntityMessenger.messageSelfPeriodically()` should be called from the `createNew()` / `loadExisting()` method or it can be called within the `createActiveEntity()` method, when the entity is not yet constructed ?
